ctci/chapter2/2.5.py

```diff
@@ -0,0 +1,87 @@
+#2.5
+#Sum Lists: You have two numbers represented by a linked list,
+#where each node contains a single digit. The digits are stored
+#in reverse order, such that the 1's digit is at the head of the list. 
+#Write a function that adds the two numbers and returns the sum as a linked list.
+
+#assuming no cycles
+
+from random import randrange
+
+class Node:
+    def __init__(self, dataval=None):
+        self.dataval = dataval
+        self.nextval = None
+
+class SLinkedList:
+    def __init__(self):
+        self.headval = None
+    
+    @classmethod
+    def genList(cls, len :int, max: int):
+        l = SLinkedList()
+        l.headval = Node(randrange(max))
+        it = l.headval
+
+        for i in range(len):
+            n = Node (randrange(max))
+            it.nextval = n
+            it = it.nextval
+        return l
+
+    def printList(self):
+        s = ""
+        it = self.headval
+        while (it != None):
+            s += str((it.dataval)) + ", "
+            it = it.nextval
+        print(s)
+
+
+def addList(l1:SLinkedList, l2:SLinkedList) -> SLinkedList:
+    l = SLinkedList()
+    i = l1.headval
+    j = l2.headval
+
+    li = Node()
+    l.headval = li
+    carry = 0
+    while (i != None or j != None or carry != 0):
+        n = Node(None)
+        if (i == None and j == None):
```

Comment:
:dizzy_face::dizzy_face::dizzy_face::dizzy_face:

One idea to clean this up is to spread it across several while loops

```python
l == SLinkedList()
l.headval = Node(0) // add a dummy value to l so dont have to worry about if l == null, just return l.nextval
while(both not null):
    n = Node(None)
    n.dataval = (j.dataval + i.dataval + carry) % 10
    carry = (j.dataval + i.dataval + carry) // 10
while(i not null): // ie j is null
    n.dataval = (j.dataval + carry) % 10
     carry = (j.dataval + carry) // 10
....
```

Doesn't matter much now, but for future problems its something to think about :smile: